plugins/nf-lamin/src/main/nextflow/lamin/LaminObserver.groovy

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright 2025, Lamin Labs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package nextflow.lamin
+
+import java.nio.file.Path
+import java.nio.file.PathMatcher
+import java.util.concurrent.locks.Lock
+import java.util.concurrent.locks.ReentrantLock
+
+import groovy.transform.CompileStatic
+import groovy.util.logging.Slf4j
+
+import nextflow.Session
+import nextflow.processor.TaskHandler
+import nextflow.processor.TaskRun
+import nextflow.trace.TraceObserver
+import nextflow.trace.TraceRecord
+
+import nextflow.lamin.api.Helper
+
+/**
+ * Example workflow events observer
+ *
+ * @author Robrecht Cannoodt <robrecht@data-intuitive.com>
+ */
+@Slf4j
+@CompileStatic
+class LaminObserver implements TraceObserver {
+    private Session session
+
+    private List<PathMatcher> matchers = []
+
+    private Set<TaskRun> tasks = []
+
+    private Set<Path> workflowInputs = []
+    private Map<Path,Path> workflowOutputs = [:]
+
+    private Lock lock = new ReentrantLock()
+
+    void logInfo(String message) {
+        log.info("nf-lamin> " + message)
+    }
+
+    @Override
+    void onFlowCreate(Session session) {
+        // store the session for later use
+        this.session = session
+
+        logInfo "onFlowCreate triggered!"
+
+        Helper.test()
+
+        def wfMetadata = session.getWorkflowMetadata()
+        def key = wfMetadata.scriptFile.toString().replaceFirst("${wfMetadata.projectDir}/", "")
+
+        // session.baseDir + "/" + session.scriptName
+
+        def description = session.config.navigate("manifest.description") as String
+
+        logInfo "Fetch or create Transform object:\n" +
+            "  trafo = ln.Transform(\n" +
+            "    key=\"${key}\",\n" +
+            "    version=\"${wfMetadata.revision}\",\n" +
```

Comment:
I think you will need to add a "source_code" otherwise you cannot version pipelines in Lamin. The source_code defines the Transform hash.

Reply:
Which `source_code` did you add, Julianus? I wanted to remove the need for providing source code for abstract pipelines even though in the case of nextflow pipelines, who I understand always have their own git repo, adding the hash of the repo state or a version number of that repo would be a legit proxy.

Reply:
Yes, that is exactly what I did. It is basically the url to the GitHub repo with the commitID or revision.

Below is the code that I am currently using.

```python

    hub = "https://github.com"
    reference = f"{hub}/{name}"

    reference_w_version = f"{reference}/tree/{commit}" if commit else f"{reference}/tree/{revision}"
    if description is None:
        description = f"Nextflow pipeline {pipeline_name}"

    t = ln.Transform(
        key=name,
        description=description,
        version=pipeline_version,
        source_code=reference_w_version,
        type="pipeline",
        reference=reference,
        reference_type="url",
    )
```

Reply:
Fantastic, thanks! 👍 

I will also need to append the mainScript to this because not every workflow is in the root of the repository.

Reply:
I moved this conversation to https://github.com/laminlabs/nf-lamin/issues/57